# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def addTwo(x):
    return x + 2

print(addTwo(4))

# The Lambda version...
# Without using a variable to pass in data
print((lambda x: x + 2)(4))

# Using a variable
a_num = 4
print((lambda x: x + 2)(a_num))

6
6
6


#### Saving to a Variable

In [4]:
f_test = lambda x: x + 2

f_test(4)

6

#### Multiple Inputs

In [11]:
# Multiple inputs with no variables
print((lambda x, y, z: x * y * z)(3, 5, 8))

# Multiple Inputs with a variable attached
x_test = lambda x, y, z: x * y * z
print(x_test(3,5,8))

120
120


#### Passing a Lambda into a Function

In [12]:
def multiply(f, num):
    '''
        f expects a lambda function
        num expects a number
    '''
    return f(num) # Up to here this is a normal function

# Below is where we call the lambda function

multiply(lambda x: x * x, 4)

16

#### Returning a Lambda from a Function

In [20]:
# Regular defined function
def multiply_tst(num):
    return num + 4

# Function within a function:
def returnFunc():
    def multiply(num):
        return num * 2
    return multiply

f_return = returnFunc()
# print(returnFunc())
print(f_return(4))

# Lambda function returned from a regular finction
def returnLam(b,c):
    return lambda x, a: x + a + b + c

r_lamb = returnLam(4,5)
print(r_lamb(5,5))

8
19


#### If Statements within Lambdas

In [21]:
# Lambda x: True if (condition) else False
f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8))
print(f_condition(12))
print(f_condition(10))

10
24
12


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [38]:
f = lambda x: x ** 2

print(f(4))



16


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [23]:
# map(function, iterable)
# An iterable has to be something like a list, dict, tuple or set
# Normally the usage of map happens with a pre-defined function - but we can use lambdas as well

def squared(num, num2):
    if num < 10 and num2 < 10:
        return num ** 2, num2 **2
    else:
        return num, num2
    
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_map = list(map(squared, numbers, more_nums)) #its important to use list() to be able to print the 'map'
print(squared_nums_map)

# map work very much like zip() as one of the exercises i did in the past combining two separate lists of strings
# NOTE : Even if there are not enough objects in both lists, being 6 objects in 'numbers' and 5 in 'more_nums'
# map() will pair whatever it can and run the function or instructions therein. 

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### Using Lambda's with Map

In [26]:
# map(lambda x: do something, iterable)
# Using lambda in map happens inline or on one line
# list(map(lambda x: x * 2, nums))

    
numbers = [4, 11, 20, 3, 15, 20]
more_nums = [4, 10, 3, 2, 6]

squared_nums_lambda = list(map(lambda x, y: (x ** 2, y ** 2) if x < 10 and y < 10 else (x,y), numbers, more_nums))
print(squared_nums_lambda)

print(list(map(lambda x, y: (x ** 2, y ** 2) if x < 10 and y < 10 else (x,y), numbers, more_nums)))

# We got rid of the extra variable 'squared_nums_lambda' as above and print it in one line

# /SAME EXERCISE AS THE ONE ABOVE BUT IN ONE LINE

[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]
[(16, 16), (11, 10), (20, 3), (9, 4), (15, 6)]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [30]:
print(list(map(lambda x, y: ((x * 2) - 1, (y * 2) - 1), numbers, more_nums)))

[(7, 7), (21, 19), (39, 5), (5, 3), (29, 11)]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [44]:
names= ['Bob', 'Andy', 'Max', 'Evan', 'Angelica']

def a_names(name):
    if name[0].lower() == 'a':
        return True
    elif name[0].lower() == 'b':
        return True
    else:
        return False

# RETURNING TRUE OR FALSE IS VERY IMPORTANT
# If it returns True it will added to our list of filtered results,(In this case is the list named
# 'new_names_list'), and if False , it will not add it to our list
    
new_names_list = list(filter(a_names, names))
print(new_names_list)


    

['Bob', 'Andy', 'Angelica']


#### Using Lambda's with Filter()

In [54]:
new_names_lambda = list(filter(lambda name: True if name[0].lower() == 'a' else False, names))
print(new_names_lambda)

['Andy', 'Angelica']


#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [75]:
from statistics import mean

numbers = [4, 11, 25, 3, 15, 20]

mean_numbers = mean(numbers)

print(mean_numbers)

print('---------Exercise Answer :---------')

mean_nums = list(filter(lambda x: True if x < mean(numbers) else False , numbers ))
print(mean_nums)


13
---------Exercise Answer :---------
[4, 11, 3]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [91]:
from functools import reduce

# reduce(function, iterable[,initialier]) #initializer is optional . if you add eg. 100 as an initializer
                                          #reduce will start the computation at 100

list_1 = [2, 4, 6, 8, 10]

def addNums(num1, num2):
    return num1 + num2

result_add = reduce(addNums, list_1)

print(result_add)

# Substract a list of numbers
def subtractNums(num1, num2):
    return num1 - num2

result_sub = reduce(subtractNums, list_1)
print(result_sub)
print(type(list_1))

#  the regular for - loop allternative :
total = 0
for i in list_1:
    total = total + i #(total += i)
print(total)

30
-26
<class 'list'>
30


#### Using Lambda's with Reduce()

In [97]:
result_lamb = reduce(lambda x, y: x + y, list_1 )
print(result_lamb)

result_sub_lamb = reduce(lambda x, y: x - y, list_1)
print(result_sub_lamb)

30
-26


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [99]:
my_list = [1, 2, 3, 4]

my_res = reduce(lambda x, y: x * y, my_list)
print(my_res)

24


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [106]:
# Recusrsion is calling the function repeatedly within the same function as it is in the 'return' on line 26

####  return 'variable or value' 'operation' '(function_name)again(variable - the step (in this case 1))'

# base case is the last number....keep reading for the example to be more clear

def addNums(num):
    #Here is our base caswe...
    if num <= 1:
        print('addNums(1) = 1')
        return num
    else:
        print(f'addNums({num}) = {num} + addNums({num - 1})')

# If we print now ....
addNums(5)

# ADDING RECURSIO FACTOR ....
# Notice the difference on the 'else' part

def addNums1(num1):
    #Here is our base caswe...
    if num1 <= 1:                #If we dont add this condition, which is the 'base case' and tells the loop
        print('addNums(1) = 1')  # to stop, it will continue printing for ever into the negative numbers in this
        return num1              #particular case
    else:
        print(f'addNums({num1}) = {num1} + addNums({num1 - 1})')
        ##Recursion below
        return num1 + addNums1(num1 - 1) #addNums(num1 - 4) is calling itslef in the return with the less value
                                          # in this case in the second loop would be 4, then 3 and so on...while
                                         # adding the results num1(5) plus addNums(num1 - 4) which is 5 - 4 and 
                                          #so on....

# If we print now ....
addNums1(5) #This is only printing the if-else loop results
print(addNums1(5)) #result of the addition (5 + 4+ 3 + 2 + 1) = 15

addNums(5) = 5 + addNums(4)
addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1
addNums(5) = 5 + addNums(4)
addNums(4) = 4 + addNums(3)
addNums(3) = 3 + addNums(2)
addNums(2) = 2 + addNums(1)
addNums(1) = 1
15


#### Writing a Factorial Function

In [115]:
# 5! = 5 * 4 * 3 * 2 * 1

def factorial(num5):
    if num5 <= 1:
        print(f'Num is {num5} ')
        return 1
    else:
        return num5 * factorial(num5 - 1)
    
factorial(5)
print(factorial(5))
    

Num is 1 
Num is 1 
120


#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [136]:
# return num -
# num8 = 0
def subRecur(num7):
    if num7 <= 1:
        return 1
    else:
        
        return num7 - subRecur(num7 - 1)

subRecur(5)
# print(subRecur(5))
    

3

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [137]:
def my_range(stop, start, step=2):
    while start < stop:
        yield start #yield key word denotes a generator
        start += step
        
for i in my_range(20, start=2):    #This i scalling the function my_range with the parameters denoted
    my_generator_value = i         # stop = 20 and start = 2. step = 2 is already set in the function my_range.
    print(my_generator_value)

2
4
6
8
10
12
14
16
18


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

# As in the exercise above, just getting rid of :
# while start < stop:
# will not have the stop = 20 as a stop signal compared to 'start'

# #NOT GOOD

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [140]:
def class_ex(stop, start, step=1):
    while start > stop:
        yield start ** 2
        start -= step
        
for i in class_ex(1, start = 10):
    squared_n = i
    print(squared_n)

100
81
64
49
36
25
16
9
4


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [175]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]

new_list = list((i.strip() for i in places))


no_spaces = list(filter(lambda x: True if x != "" else False, new_list))
print(no_spaces)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [353]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author1 = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
# print(author)

# Lowering all cases

new_list5 = list(i.lower() for i in author)
# print(new_list5)

# spliting the strings
new_list6 = list(x.split() for x in new_list5)
# print(new_list6)

# reversing the order of each sub-list
new_list7 = list(x[::-1] for x in new_list6)
# print(new_list7)

# title() for each proper name
new_list8 = list(list(f.title() for f in g) for g in new_list7)
# print(new_list8)

# sorting the list alphabetically based on last names
new_list8.sort()

# reversing names back to original form
new_list9 = list(x[::-1] for x in new_list8)
# print(new_list9)

# joining each sub-string
final_list = list((' '.join(str(x) for x in g)) for g in new_list9)

print(final_list)

# author.sort(key= lambda x: x.lower().split()[-1])

# print(author)

author.sort(key=lambda names: names.lower().split()[-1])



print(author)


['Victor Anisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David Hasselhoff']
['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [375]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

def cel_to_fa(places):
    return places[0], (((9/5)*places[1]) + 32)
  
final_result = list(map(cel_to_fa, places)) #its important to use list() to be able to print the 'map'
print(final_result)

# OR

print(list(map(lambda w: [w[0], w[1] * (9/5) + 32], places)))

# final_result1 = list(map((lambda x: (((9/5)*places[1]) + 32) for x in places)))
# print(final_result1)
# --------------------------------------------------


    


[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
[['Nashua', 89.6], ['Boston', 53.6], ['Los Angelos', 111.2], ['Miami', 84.2]]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [373]:
def fib_seq(rec):
   if rec <= 1:
       return rec
   return fib_seq(rec-1) + fib_seq(rec-2)

num_rec = int(input('Please enter the number of recursions? '))

print(f'Output for fib({num_rec})')
for i in range(num_rec):
    print(f'Iteration {i}: ', fib_seq(i+1))

Please enter the number of recursions? 5
Output for fib(5)
Iteration 0:  1
Iteration 1:  1
Iteration 2:  2
Iteration 3:  3
Iteration 4:  5
